### PR TITLE
[RLlib; Offline RL] Store episodes in state form.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

Storing with `ray.data` episodes as instances results in pickled instances that are maybe not compatible with later python versions. This PR tries to develop a stream that is compatible with later Python versions by simplifying objects in their `get_state` methods and tries to reproduce them in their static `from_state` method.
It uses `mgspack` to serialize objects (and `msgpack-numpy` to serialize numpy array) which is a serialization protocol independent of Python versions. 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/env/single_agent_episode.py [hunk 1]
```diff
@@ -1695,32 +1696,44 @@ def get_state(self) -> Dict[str, Any]:
...
             "multi_agent_episode_id": self.multi_agent_episode_id,
-            # TODO (simon): Check, if we need to have a `get_state` method for
-            #  `InfiniteLookbackBuffer` and call it here.
-            "observations": self.observations,
-            "actions": self.actions,
-            "rewards": self.rewards,
-            "infos": self.infos,
-            "extra_model_outputs": self.extra_model_outputs,
+            # Note, all data is stored in `InfiniteLookbackBuffer`s.
+            "observations": self.observations.get_state(),
```
Comment (sven1977) [inline rllib/env/single_agent_episode.py:1712]: Nice!

File: rllib/env/single_agent_episode.py [hunk 2]
```diff
@@ -1695,32 +1696,44 @@ def get_state(self) -> Dict[str, Any]:
...
             "t_started": self.t_started,
             "t": self.t,
-            "_observation_space": self._observation_space,
-            "_action_space": self._action_space,
+            "_observation_space": gym_space_to_dict(self._observation_space)
+            if self._observation_space
+            else self._observation_space,
+            "_action_space": gym_space_to_dict(self._action_space)
+            if self._action_space
+            else self._action_space,
```
Comment (sven1977) [inline rllib/env/single_agent_episode.py:1731]: nit: "else None" would be better, imo.

Same above for obs space.

File: rllib/env/single_agent_episode.py [hunk 3]
```diff
@@ -1695,32 +1696,44 @@ def get_state(self) -> Dict[str, Any]:
...
             "_start_time": self._start_time,
             "_last_step_time": self._last_step_time,
-            "_temporary_timestep_data": self._temporary_timestep_data,
+            "_temporary_timestep_data": dict(self._temporary_timestep_data)
```
Comment (sven1977) [inline rllib/env/single_agent_episode.py:1734]: cool!

File: rllib/env/single_agent_episode.py [hunk 4]
```diff
@@ -1739,21 +1752,48 @@ def from_state(state: Dict[str, Any]) -> "SingleAgentEpisode":
...
         episode.module_id = state["module_id"]
         episode.multi_agent_episode_id = state["multi_agent_episode_id"]
-        episode.observations = state["observations"]
-        episode.actions = state["actions"]
-        episode.rewards = state["rewards"]
-        episode.infos = state["infos"]
-        episode.extra_model_outputs = state["extra_model_outputs"]
+        # Convert data back to `InfiniteLookbackBuffer`s.
+        episode.observations = InfiniteLookbackBuffer.from_state(state["observations"])
+        episode.actions = InfiniteLookbackBuffer.from_state(state["actions"])
```
Comment (sven1977) [inline rllib/env/single_agent_episode.py:1757]: nice!!

File: rllib/env/single_agent_episode.py [hunk 5]
```diff
@@ -1739,21 +1752,48 @@ def from_state(state: Dict[str, Any]) -> "SingleAgentEpisode":
...
         episode.is_truncated = state["is_truncated"]
         episode.t_started = state["t_started"]
         episode.t = state["t"]
-        episode._observation_space = state["_observation_space"]
-        episode._action_space = state["_action_space"]
+        # We need to convert the spaces to dictionaries for serialization.
+        episode._observation_space = (
+            gym_space_from_dict(state["_observation_space"])
+            if state["_observation_space"]
+            else state["_observation_space"]
```
Comment (sven1977) [inline rllib/env/single_agent_episode.py:1785]: nit: "else None"?

File: rllib/env/single_agent_episode.py [hunk 6]
```diff
@@ -1739,21 +1752,48 @@ def from_state(state: Dict[str, Any]) -> "SingleAgentEpisode":
...
+        )
+        episode._action_space = (
+            gym_space_from_dict(state["_action_space"])
+            if state["_action_space"]
+            else state["_action_space"]
```
Comment (sven1977) [inline rllib/env/single_agent_episode.py:1790]: nit: "else None"

File: rllib/env/utils/infinite_lookback_buffer.py [hunk 7]
```diff
@@ -34,6 +35,77 @@ def __init__(
         self.space_struct = None
         self.space = space
 
+    def __eq__(
```
Comment (sven1977) [inline rllib/env/utils/infinite_lookback_buffer.py:38]: dumb question: why do we need this method?

**Review by sven1977 — APPROVED** [review]
Great PR. Thanks @simonsays1980 . Just one question on __eq__ and a handful of nits.

File: rllib/env/utils/infinite_lookback_buffer.py (hunk 7 above)
Comment (simonsays1980) [reply rllib/env/utils/infinite_lookback_buffer.py:38]: Just for testing purposes to make comparisons of instances before writing and after writing.

**Pull request merged**